File: backend/layers/common/nodejs/src/auth-utils.ts
import type { APIGatewayProxyEventV2 } from './types/api';
import { verifyAccessToken, type AccessTokenPayload } from './jwt';

export interface AuthContext {
  userId: string;
  role: string;
  email?: string;
}

export function extractTokenFromEvent(
  event: APIGatewayProxyEventV2
): string | null {
  const authHeader =
    event.headers?.authorization || event.headers?.Authorization;
  if (!authHeader) {
    return null;
  }

  if (authHeader.startsWith('Bearer ')) {
    return authHeader.substring(7).trim();
  }

  return null;
}

export function getUserFromEvent(
  event: APIGatewayProxyEventV2
): AuthContext | null {
  // First check if the authorizer already decoded the token (set by Lambda authorizer)
  const requestContext = event.requestContext as Record<string, unknown>;
  const authorizer = requestContext?.authorizer as
    | Record<string, unknown>
    | undefined;
  const lambda = authorizer?.lambda as Record<string, string> | undefined;

  if (lambda?.userId) {
    return {
      userId: lambda.userId,
      role: lambda.role || 'student',
      email: lambda.email,
    };
  }

  // Fallback: decode the token directly
  const token = extractTokenFromEvent(event);
  if (!token) {
    return null;
  }

  try {
    const decoded: AccessTokenPayload = verifyAccessToken(token);
    return {
      userId: decoded.userId,
      role: decoded.role,
    };
  } catch {
    return null;
  }
}

export function requireAuth(event: APIGatewayProxyEventV2): AuthContext {
  const user = getUserFromEvent(event);
  if (!user) {
    throw new Error('Unauthorized');
  }
  return user;
}

export function requireAdmin(event: APIGatewayProxyEventV2): AuthContext {
  const user = requireAuth(event);
  if (user.role !== 'admin') {
    throw new Error('Forbidden: admin access required');
  }
  return user;
}

File: backend/layers/common/python/auth_utils.py
from jwt_utils import verify_access_token

def extract_token(event):
    auth_header = event.get('headers', {}).get('authorization', '') or event.get('headers', {}).get('Authorization', '')
    if auth_header.startswith('Bearer '):
        return auth_header[7:]
    return None

def get_user_from_event(event):
    token = extract_token(event)
    if not token:
        return None
    return verify_access_token(token)

def require_auth(event):
    user = get_user_from_event(event)
    if not user:
        return None, {'statusCode': 401, 'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}, 'body': '{"error":"Unauthorized"}'}
    return user, None

def require_admin(event):
    user, err = require_auth(event)
    if err:
        return None, err
    if user.get('role') != 'admin':
        return None, {'statusCode': 403, 'headers': {'Content-Type': 'application/json', 'Access-Control-Allow-Origin': '*'}, 'body': '{"error":"Admin access required"}'}
    return user, None
